scipy/optimize/tests/test_trustregion_exact.py

```diff
@@ -0,0 +1,360 @@
+"""
+Unit tests for trust-region iterative subproblem.
+
+To run it in its simplest form::
+  nosetests test_optimize.py
+
+"""
+from __future__ import division, print_function, absolute_import
+
+import numpy as np
+from scipy.optimize._trustregion_exact \
+    import (estimate_smallest_singular_value,
+            singular_leading_submatrix,
+            IterativeSubproblem)
+from scipy.linalg import (svd, get_lapack_funcs, det,
+                          cho_factor, cho_solve, qr,
+                          eigvalsh, eig, norm)
+from numpy.testing import (TestCase, assert_, assert_array_equal,
+                           assert_equal, assert_array_almost_equal,
+                           assert_array_less, run_module_suite)
+
+
+def random_entry(n, min_eig, max_eig, case):
+
+    # Generate random matrix
+    rand = np.random.uniform(-1, 1, (n, n))
+
+    # QR decomposition
+    Q, _, _ = qr(rand, pivoting='True')
+
+    # Generate random eigenvalues
+    eigvalues = np.random.uniform(min_eig, max_eig, n)
+    eigvalues = np.sort(eigvalues)[::-1]
+
+    # Generate matrix
+    Qaux = np.multiply(eigvalues, Q)
+    A = np.dot(Qaux, Q.T)
+
+    # Generate gradient vector accordingly
+    # to the case is being tested
+    if case == 'hard':
+        g = np.zeros(n)
+        g[:-1] = np.random.uniform(-1, 1, n-1)
+        g = np.dot(Q, g)
+    elif case == 'jac_equal_zero':
+        g = np.zeros(n)
+    else:
+        g = np.random.uniform(-1, 1, n)
+
+    return A, g
+
+
+class TestEstimateSmallestSingularValue(TestCase):
```

Comment:
About `TestCase`. `TestCase` is a class from `unittest` Python standard module. It doesn't play in any way with `nosetests` used for `scipy`.

So it's usage seems dubious, people from `scikit-learn` didn't allow me to use it. So you can:

a. Keep everything as it is now.
b. Use only functions (scikit-learn approach).
c. Don't inherit from `TestCase` (for clarity), but group tests in classes as now for convenience.

Just something to consider, I don't mind either way.

For this particular test group, I don't see any value using a class.